rclcpp_action/src/server.cpp

```diff
@@ -627,19 +627,19 @@ ServerBase::publish_result(const GoalUUID & uuid, std::shared_ptr<void> result_m
   }
 
   {
-    std::lock_guard<std::recursive_mutex> lock(pimpl_->unordered_map_mutex_);
+    std::lock_guard<std::recursive_mutex> unordered_map_lock(pimpl_->unordered_map_mutex_);
```

Comment:
This seems fine to me, but it can introduce a ptential deadlock in the future.
I would rather:

- Document in which order we're currently taking the `unordered_map_mutex_`, `action_server_reentrant_mutex_` mutexes, to avoid someone doing it in the opposite order later.
- Take `action_server_reentrant_mutex_` out of the if/loop together with  `unordered_map_mutex_` using `std::scoped_lock`.

Reply:
With using `std::scoped_lock.`, then if could not be backported to `foxy`? Because `foxy` is compiled with C++14.

Reply:
Sounds good, but please add documentation explaining what's the current order we're taking the mutexes.

Reply:
I added a NOTE comment to this block scope.